tests/integration_tests/data_steward/cdr_cleaner/cleaning_rules/ensure_date_datetime_consistency_test.py

```diff
@@ -0,0 +1,393 @@
+
+        sb_table_names = cls.query_class.get_sandbox_tablenames()
+        for table_name in sb_table_names:
+            cls.fq_sandbox_table_names.append(
+                f'{project_id}.{sandbox_id}.{table_name}')
+
+        for key in TABLE_DATES:
+            cls.fq_table_names.append(f"{project_id}.{dataset_id}.{key}")
+
+        # call super to set up the client, create datasets, and create
+        # empty test tables
+        # NOTE:  does not create empty sandbox tables.
+        super().setUpClass()
+
+    def setUp(self):
+        """
+        Add data to the tables for the rule to run on.
+        """
+        load_statements = []
+        # create the string(s) to load the data
+        fq_dataset_name = self.fq_table_names[0].split('.')
+        self.fq_dataset_name = '.'.join(fq_dataset_name[:-1])
+        for tmpl in self.insert_fake_participants_tmpls:
+            query = tmpl.render(fq_dataset_name=self.fq_dataset_name)
+            load_statements.append(query)
+
+        self.load_test_data(load_statements)
+
+    def test(self):
+        """
+        Use the default drop rows test function.
+
+        Validates pre-conditions, test execution and post conditions based on
+        the load statements and the tables_and_counts variable.
+        """
+        start_date = parser.parse('2016-05-01').date()
+        start_datetime = parser.parse('2016-05-01 11:00:00 UTC')
+        start_default_datetime = parser.parse('2016-05-01 00:00:00 UTC')
+        end_date = parser.parse('2016-05-02').date()
+        end_datetime = parser.parse('2016-05-02 11:00:00 UTC')
+        end_default_datetime = parser.parse('2016-05-02 00:00:00 UTC')
+
+        tables_and_counts = [{
+            'fq_table_name':
+                '.'.join([self.fq_dataset_name, 'specimen']),
+            'fq_sandbox_table_name':
+                '',
+            'loaded_ids': [701, 702, 703, 704],
+            'sandboxed_ids': [],
+            'fields': ['specimen_id', 'specimen_date', 'specimen_datetime'],
+            'cleaned_values': [(701, start_date, start_default_datetime),
+                               (702, start_date, start_datetime),
+                               (703, start_date, start_datetime),
+                               (704, start_date, start_datetime)]
+        }, {
+            'fq_table_name':
+                '.'.join([self.fq_dataset_name, 'measurement']),
+            'fq_sandbox_table_name':
+                '',
+            'loaded_ids': [301, 302, 303, 304],
+            'sandboxed_ids': [],
+            'fields': [
+                'measurement_id', 'measurement_date', 'measurement_datetime'
+            ],
+            'cleaned_values': [(301, start_date, start_default_datetime),
+                               (302, start_date, start_datetime),
+                               (303, start_date, start_datetime),
+                               (304, start_date, start_datetime)]
+        }, {
+            'fq_table_name':
```

Comment:
Since we're trying to establish a convention here, a few higher level comments:

- is it important for the purposes of this integration test to check the behavior on every affected table? Or would a subset suffice? Checking all tables feels to me a bit like it's verging on a change detector test, e.g. if we forgot to apply this transform to a particular table, we're equally likely to have forgotten it in the test. When this test breaks, it's likely only because we've updated the set of tables for which we're applying this transform.
- The association between data and test might be more clearly expressed here with individual test cases to check specific scenarios, right now it's difficult to correlate these expectations back to the original data, and understand intent with the assertion.
- alternatively, another way to write this test more clearly could be to extend these `table_infos` to include `initial_values` inline. Then the base test would insert them all before running the query. You might run into problems with cross-references in this model though.

Reply:
I am most interested in discussing your first bullet point.  I had initially only tested a single table.  I went ahead and tested the other tables for completeness sake.  Doing so, I discovered the schema differences required loading different dataset types.  We could still possibly do a subset, but I wouldn't go with a single table because this rule needs to be applied to:
1. multiple field sets on some tables and single field sets on others
2. required vs nullable fields

Reply:
Makes sense - I would suggest perhaps thinning this down to just a representative table for each of the cases you described (and making it clear what is being tested for in each case).

Reply:
I added test cases for two scenarios.  I want to keep the 'initial_values` out of the `table_and_counts` elements.  In lining this could complicate loading and reading fields in this data structure.